### DISCUSSION TITLE
How to speed up wasm build times for faster development iteration?

### DISCUSSION THREAD

**Discussion by bluenote10** [gfx-rs/wgpu#3654] [Q&A] [answered]
I'm new to both wgpu and targeting wasm, so I may simply lack some basic tricks.

I have started with a basic variation of the [hello-triangle](https://github.com/gfx-rs/wgpu/tree/trunk/wgpu/examples/hello-triangle) example and building my project with `wasm-pack build --target web`. That works fine, but I noticed that my development iteration cycles are quite slow: Re-compiling my wasm target takes roughly half a minute. It looks like `wasm-opt` is the dominating contribution to the re-build time. In some other wasm projects of mine that don't use wgpu the re-build times are actually pretty fast.

I'm wondering what are the general recommendations for fast development iteration cycles for wgpu + wasm? Is it e.g. possible to turn off wasm-opt, or is there a kind of "fast optimization" profile? If so, will it only impact the wasm size, or is the optimization actually need for runtime performance as well?

**Comment by daxpedda** [comment]
Personally I don't have any experience with `wasm-pack`, but in the past it definitely supported turning off `wasm-opt`.
`wasm-opt` definitely impacts performance, but it shouldn't be significant for development builds.

Without `wasm-opt`, development iteration cycles should be pretty short, but I never figured out a way to decrease them further.

**Comment by kpreid** [comment, marked as answer]
`wasm-pack` [defaults to](https://rustwasm.github.io/docs/wasm-pack/commands/build.html#profile) a *release* build (optimized by `rustc`, like `cargo build --release`) in addition to running `wasm-opt`.

You can change both of these at once by running `wasm-pack build --dev --target web`. It will use [the cargo `dev`  profile](https://doc.rust-lang.org/cargo/reference/profiles.html) and will not run `wasm-opt`.

Also, if you want to disable `wasm-opt` even for releases, you can [configure it using your `Cargo.toml`](https://rustwasm.github.io/docs/wasm-pack/cargo-toml-configuration.html).
